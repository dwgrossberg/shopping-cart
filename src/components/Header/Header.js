import React from "react";
import { NavLink } from "react-router-dom";
import icon from "../../assets/icon.png";
import Cart from "../Cart/Cart";
import "./Header.scss";

const Header = (props) => {
  return (
    <div className="header">
      <nav className="header-nav">
        <ul>
          <li>
            <img alt="icon" src={icon} id="icon"></img>
          </li>
          <NavLink
            to="/"
            className={"nav-link"}
            style={{ textDecoration: "none" }}
          >
            <li>home</li>
          </NavLink>
          <NavLink to="/products" className={"nav-link"}>
            <li>products</li>
          </NavLink>
          <NavLink to="/contact" className={"nav-link"}>
            <li>contact</li>
          </NavLink>
        </ul>
      </nav>
      <Cart />
    </div>
  );
};

export default Header;
