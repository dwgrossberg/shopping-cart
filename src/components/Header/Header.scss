.header {
  position: fixed;
  display: flex;
  top: 0;
  left: 0;
  right: 0;
  padding: 20px;
  justify-content: space-between;
  background-color: #ede8ed;
  opacity: 0.9;
  z-index: 100;
  .header-nav ul {
    display: flex;
    align-items: center;
    gap: 10px;
    list-style: none;
    .nav-link {
      border-bottom: 3px solid transparent;
      padding: 0 0 2px;
      transition: border-bottom 0.75s ease;
      border-radius: 1px;
      color: #820b8a;
      &:hover {
        border-bottom: 3px solid #820b8a;
      }
    }
  }
}

.header-nav {
  display: flex;
  align-items: center;
  li {
    display: flex;
    align-items: center;
  }
}

#icon {
  height: 20px;
}
