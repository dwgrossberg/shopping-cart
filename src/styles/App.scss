.App {
  height: 100%;
  min-height: (100vh);
  width: 100%;
  gap: 20px;
  display: grid;
  align-items: stretch;
  justify-items: center;
  grid-template: 1fr 50px / 1fr;
  padding: 30px;
  background-color: ghostwhite;
}

.price-quant {
  display: flex;
  align-items: center;
  justify-content: space-around;
  width: 80%;
  gap: 20px;
  .quant-number {
    width: 10px;
    text-align: center;
    font-size: 9px;
  }
  .quant {
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 10px;
    button {
      width: 18px;
      height: 18px;
      border-radius: 7px;
      background-color: white;
      border: 0.5px solid #af9ab2;
      cursor: pointer;
      &:hover {
        background-color: #672a4e;
        color: white;
      }
    }
  }
}
