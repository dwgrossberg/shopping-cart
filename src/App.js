import React, { useState, useEffect } from "react";
import { BrowserRouter, Routes, Route } from "react-router-dom";
import Header from "./components/Header/Header";
import Footer from "./components/Footer/Footer";
import Home from "./pages/home/home";
import Products from "./pages/products/products";
import Contact from "./pages/contact/contact";
import "./styles/App.scss";
import ScrollToTop from "./components/ScrollToTop";

const App = () => {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useState([]);
  const [category, setCategory] = useState([]);

  // shuffle array to randomize product display on each visit
  const shuffle = (array) => {
    let currentIndex = array.length;
    let randomIndex;
    while (currentIndex !== 0) {
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex--;
      [array[currentIndex], array[randomIndex]] = [
        array[randomIndex],
        array[currentIndex],
      ];
    }
    return array;
  };

  // fetch product data from fake store api
  const fetchProductData = async () => {
    try {
      const prodResponse = await fetch("https://fakestoreapi.com/products");
      const prodData = await prodResponse.json();
      const shuffledData = shuffle(prodData);
      setProducts(shuffledData);
      setCategory(shuffledData);
      console.log(prodData);
      return prodData;
    } catch (err) {
      console.log(err);
    }
  };

  // fetch cart data from fake store api
  const fetchCartData = async () => {
    const randomNum = Math.floor(Math.random() * 6 + 1);
    try {
      const cartResponse = await fetch(
        `https://fakestoreapi.com/carts/${randomNum}`
      );
      const cartData = await cartResponse.json();
      setCart(cartData.products);
      console.log(cartData.products);
      return cartData;
    } catch (err) {
      console.log(err);
    }
  };

  useEffect(() => {
    fetchProductData();
    fetchCartData();
  }, []);

  return (
    <div className="App">
      <BrowserRouter>
        <Header cart={cart} />
        <ScrollToTop />
        <Routes>
          <Route path="/" element={<Home />} />
          <Route
            path="/products"
            element={
              <Products
                products={products}
                setProducts={setProducts}
                category={category}
                setCategory={setCategory}
                cart={cart}
                setCart={setCart}
                shuffle={shuffle}
              />
            }
          />
          <Route path="/contact" element={<Contact />} />
        </Routes>
        <Footer />
      </BrowserRouter>
    </div>
  );
};

export default App;
