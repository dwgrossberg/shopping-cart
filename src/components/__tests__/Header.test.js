import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import { screen, render, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Header from "../Header/Header";
import App from "../../App";

describe("Header component", () => {
  afterEach(cleanup);

  Object.defineProperty(window, "scrollTo", {
    value: () => {},
    writable: true,
  });

  it("snapshot test", () => {
    const renderer = new ShallowRenderer();
    renderer.render(<Header />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders correct nav links", () => {
    render(<App />);
    expect(screen.getByText(/home/i)).toBeInTheDocument();
    expect(screen.getByText(/products/i)).toBeInTheDocument();
    expect(screen.getByText(/contact/i)).toBeInTheDocument();
    expect(screen.getByAltText(/shopping-cart/i)).toBeInTheDocument();
  });

  it("renders all route navigation links correctly", async () => {
    render(<App />);
    const user = userEvent.setup();
    expect(screen.getByText(/anything you want store/i)).toBeInTheDocument();
    await user.click(screen.getByText(/products/i));
    expect(screen.getByText(/filter shop items by/i)).toBeInTheDocument();
    await user.click(screen.getByText(/contact/i));
    expect(screen.getByRole("heading").textContent).toMatch(/contact/i);
    await user.click(screen.getByAltText(/shopping-cart/i));
    expect(screen.getByText(/your shopping cart/i)).toBeInTheDocument();
  });
});
