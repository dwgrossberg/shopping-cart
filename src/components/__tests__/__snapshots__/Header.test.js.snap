// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component snapshot test 1`] = `
<div
  className="header"
>
  <nav
    className="header-nav"
  >
    <ul>
      <li>
        <img
          alt="icon"
          id="icon"
          src="icon.png"
        />
      </li>
      <NavLink
        className="nav-link"
        style={
          Object {
            "textDecoration": "none",
          }
        }
        to="/"
      >
        <li>
          home
        </li>
      </NavLink>
      <NavLink
        className="nav-link"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
        to="/products"
      >
        <li>
          products
        </li>
      </NavLink>
      <NavLink
        className="nav-link"
        style={
          Object {
            "textDecoration": "none",
          }
        }
        to="/contact"
      >
        <li>
          contact
        </li>
      </NavLink>
    </ul>
  </nav>
  <NavLink
    to="/cart"
  >
    <CartIcon />
  </NavLink>
</div>
`;
